h/streamer/messages.py

```diff
@@ -140,23 +142,32 @@ def _generate_annotation_event(
     if user_nipsad and socket.authenticated_userid != annotation.userid:
         return None
 
-    notification = {"type": "annotation-notification", "options": {"action": action}}
+    # The `prepare` function sets the active registry which is an implicit
+    # dependency of some of the authorization logic used to look up annotation
+    # and group permissions.
+    with pyramid.scripting.prepare(registry=socket.registry):
+        notification = {
+            "type": "annotation-notification",
+            "options": {"action": action},
+        }
 
-    base_url = socket.registry.settings.get("h.app_url", "http://localhost:5000")
-    links_service = LinksService(base_url, socket.registry)
-    resource = AnnotationContext(annotation, group_service, links_service)
-    serialized = presenters.AnnotationJSONPresenter(
-        resource, formatters=formatters
-    ).asdict()
+        base_url = socket.registry.settings.get("h.app_url", "http://localhost:5000")
+        links_service = LinksService(base_url, socket.registry)
+        resource = AnnotationContext(annotation, group_service, links_service)
 
-    permissions = serialized.get("permissions")
-    if not _authorized_to_read(socket.effective_principals, permissions):
-        return None
+        # Check whether client is authorized to read this annotation.
+        read_principals = principals_allowed_by_permission(resource, "read")
+        if not set(read_principals).intersection(socket.effective_principals):
+            return None
+
+        serialized = presenters.AnnotationJSONPresenter(
```

Comment:
Good thinking, moving the serialization work after the permissions check ✅

Reply:
I plan to address this as part of https://github.com/hypothesis/h/pull/6214, which fixes the bigger problem that annotation delete events don't work properly at present.